### PR TITLE
Expose the historical success probability calculation itself

### PR DESCRIPTION
**Pull request by TheBlueMatt**
In 3f32f60 we exposed the
historical success probability buckets directly, with a long method
doc explaining how to use it. While this is great for logging
exactly what the internal model thinks, its also helpful to let
users know what the internal model thinks the success probability
is directly, allowing them to compare route success probabilities.

Here we do so but only for the historical tracking buckets.

This also has a few cleanups in the scoring logic. This is the first few commits from  #2176, without the major change to actually swap to different buckets.

### CODE REVIEW

**Review by alecchendev — COMMENTED** [review]
Still new to the scoring logic, but LGTM for the most part

File: lightning/src/routing/scoring.rs [hunk 1]
```diff
@@ -1702,6 +1547,166 @@ mod approx {
 	}
 }
 
+mod bucketed_history {
```
Comment (jkczyz) [inline lightning/src/routing/scoring.rs:1538]: Is having these in a submodule that useful? Note in another comment that I left, it may actually be better to get rid of `HistoricalMinMaxBuckets` in order to share `now` and `last_updated`.

File: lightning/src/routing/scoring.rs [hunk 2]
```diff
@@ -1079,26 +1093,13 @@ impl<L: Deref<Target = u64>, BRT: Deref<Target = HistoricalBucketRangeTracker>,
...
-			} else {
-				// Only use 128-bit arithmetic when multiplication will overflow to avoid 128-bit
-				// division. This branch should only be hit in fuzz testing since the amount would
-				// need to be over 2.88 million BTC in practice.
-				((amount_msat as u128) * 64 / (self.capacity_msat as u128).saturating_add(1))
-					.try_into().unwrap_or(65)
-			};
-			#[cfg(not(fuzzing))]
-			debug_assert!(payment_amt_64th_bucket <= 64);
-			if payment_amt_64th_bucket > 64 { return res; }
```
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:1082]: I don't quite follow. The early return that was removed was in `penalty_msat`, not `calculate_success_probability_times_billion`. Even if the latter returns `None`, the code in the `else` clause below is executed where previously it would not have been.

File: lightning/src/routing/scoring.rs [hunk 3]
```diff
@@ -1035,7 +929,9 @@ impl<L: Deref<Target = u64>, BRT: Deref<Target = HistoricalBucketRangeTracker>,
 	/// Returns a liquidity penalty for routing the given HTLC `amount_msat` through the channel in
 	/// this direction.
 	fn penalty_msat(&self, amount_msat: u64, score_params: &ProbabilisticScoringFeeParameters) -> u64 {
-		let max_liquidity_msat = self.max_liquidity_msat();
+		let available_capacity = self.available_capacity();
+		let max_liquidity_msat = available_capacity.saturating_sub(
+			self.decayed_offset_msat(*self.max_liquidity_offset_msat));
```
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:927]: Hmm... but this isn't a problem if it is inlined. Seems it would be better to keep the `max_liquidity_msat` calculation in one place than trying to optimize this.

File: lightning/src/routing/scoring.rs [hunk 4]
```diff
@@ -715,12 +714,9 @@ impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, T: Time> ProbabilisticScorerU
...
 						let dir_liq = liq.as_directed(source, target, 0, amt, self.decay_params);
 
-						let buckets = HistoricalMinMaxBuckets {
-							min_liquidity_offset_history: &dir_liq.min_liquidity_offset_history,
-							max_liquidity_offset_history: &dir_liq.max_liquidity_offset_history,
-						};
-						let (min_buckets, max_buckets, _) = buckets.get_decayed_buckets(now,
-							*dir_liq.last_updated, self.decay_params.historical_no_updates_half_life);
+						let (min_buckets, max_buckets, _) = dir_liq.liquidity_offset_history
+							.get_decayed_buckets(now, *dir_liq.last_updated,
```
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:718]: It's an implementation detail to compute the decay.

File: lightning/src/routing/scoring.rs [hunk 5]
```diff
@@ -673,8 +673,7 @@ struct ChannelLiquidity<T: Time> {
 struct DirectedChannelLiquidity<L: Deref<Target = u64>, BRT: Deref<Target = HistoricalBucketRangeTracker>, T: Time, U: Deref<Target = T>> {
 	min_liquidity_offset_msat: L,
 	max_liquidity_offset_msat: L,
-	min_liquidity_offset_history: BRT,
-	max_liquidity_offset_history: BRT,
+	liquidity_offset_history: HistoricalMinMaxBuckets<BRT>,
```
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:676]: I guess you can pass it to all `pub` functions instead.

File: lightning/src/routing/scoring.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:718]: I guess you can pass it to all `pub` functions instead.

File: lightning/src/routing/scoring.rs (hunk 1 above)
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:1538]: Hmm... I think it would make more sense separate out all the liquidity code and keep the historical tracking with it.

Comment (jkczyz) [reply lightning/src/routing/scoring.rs:1540]: > Aside from the core concept of "how do I convert liquidity bounds into a probability" (which I plan to move into a util method in a coming PR), they don't really share anything - they're doing very different calculation.

That's the penalty computation. But tracking and updating the current liquidity seems to have a relationship with historical liquidity in so far they are both code related to liquidity, even if the code itself doesn't overlap. I mean, structurally they belong together at very least given one is a sub-struct of the other. Seems odd to have those in separate modules.

> Personally, I find `scorer.rs` pretty hard to follow, there are too many structs flying around tracking liquidity bounds, current liquidity bounds, etc. If we can split it into separate, isolated (which requires a module in rust, module members can always read struct fields otherwise) bits it'd read much cleaner. I don't feel super strongly about a new file, but a module at least makes sense, and probably putting the log table in a new file cause its really distracting.

FWIW, when I said "liquidity code and keep the historical tracking with it", I meant move `ChannelLiquidity`, `DirectedChannelLiquidity`, and all the historical tracking structs into a separate module.

Comment (jkczyz) [reply lightning/src/routing/scoring.rs:1540]: > That's a large part of the file, though? Aside from the traits themselves and the log tables and tests of the same.

Yeah, implementation code is about 600 lines excluding log approximation (about 800 with). Only a few of the tests examine the liquidity structs directly. The rest look at `channel_penalty_msat`.

I'm fine putting that in one file / module and having `scoring.rs` use whatever interface that it exposes.

**Review by jkczyz — COMMENTED** [review]
LGTM aside from https://github.com/lightningdevkit/rust-lightning/pull/2466#discussion_r1303404347, but that can always wait on a follow-up. Feel free to squash.

File: lightning/src/routing/scoring.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply lightning/src/routing/scoring.rs:718]: SGTM

**Review by jkczyz — APPROVED** [review]
